core/src/main/java/com/linecorp/armeria/server/HttpHeaderUtil.java

```diff
@@ -70,64 +72,68 @@
+                        return proxiedAddresses;
                     }
                 }
             } else {
+                final ImmutableList.Builder<InetSocketAddress> builder = ImmutableList.builder();
                 final AsciiString name = source.header();
                 if (name.equals(HttpHeaderNames.FORWARDED)) {
-                    addr = getFirstValidAddress(
-                            headers.get(HttpHeaderNames.FORWARDED), FORWARDED_CONVERTER, filter);
+                    headers.getAll(HttpHeaderNames.FORWARDED).forEach(forwarded -> {
+                        getAllValidAddress(forwarded, FORWARDED_CONVERTER, filter, builder);
+                    });
                 } else {
-                    addr = getFirstValidAddress(headers.get(name), Function.identity(), filter);
+                    headers.getAll(name).forEach(header -> {
+                        getAllValidAddress(header, Function.identity(), filter, builder);
+                    });
                 }
-                if (addr != null) {
-                    return addr;
+
+                final ImmutableList<InetSocketAddress> addresses = builder.build();
+                if (!addresses.isEmpty()) {
+                    return ProxiedAddresses.of(addresses.get(0), addresses.subList(1, addresses.size()));
                 }
             }
         }
         // We do not apply the filter to the remote address.
-        return remoteAddress;
+        return ProxiedAddresses.of(remoteAddress);
     }
 
     @VisibleForTesting
-    @Nullable
-    static InetAddress getFirstValidAddress(@Nullable String headerValue,
-                                            Function<String, String> valueConverter,
-                                            Predicate<InetAddress> filter) {
+    static void getAllValidAddress(@Nullable String headerValue,
+                                   Function<String, String> valueConverter,
+                                   Predicate<InetAddress> filter,
+                                   ImmutableList.Builder<InetSocketAddress> accumulator) {
         if (Strings.isNullOrEmpty(headerValue)) {
-            return null;
+            return;
         }
         for (String value : CSV_SPLITTER.split(headerValue)) {
             final String v = valueConverter.apply(value);
             if (Strings.isNullOrEmpty(v)) {
                 continue;
             }
             try {
-                final InetAddress addr = createInetAddress(v);
-                if (addr != null && filter.test(addr)) {
-                    return addr;
+                final InetSocketAddress addr = createInetSocketAddress(v);
+                if (addr != null && filter.test(addr.getAddress())) {
+                    accumulator.add(addr);
                 }
             } catch (UnknownHostException e) {
                 logger.debug("Failed to resolve hostname: {}", v, e);
```

Comment:
not related this PR but I think we need to change this something like
`Failed to create an IP address:`

Reply:
I will address this comment. https://github.com/line/armeria/pull/2294#discussion_r354091569 😀